src/nvim/globals.h

```diff
@@ -140,7 +140,7 @@ EXTERN int mod_mask INIT(= 0x0);                /* current key modifiers */
  * temporarily moved.  The old position is restored with the next call to
  * update_screen().
  */
-EXTERN int cmdline_row;
+EXTERN long cmdline_row;
```

Comment:
Please note that you can't just change the type. 
You have to review every place where that variable is used to check the new type doesn't introduce problems.
For `cmdline_row`, there are 81 such usages.
Note that the fact that it compiles without errors doesn't mean anything, as many of those usages are in files where -Wconversion is not yet enabled. And, even if -Wconversion was enabled for all files, you still should manually review every usage, because of possible previous explicit casts becoming incorrect (those would not be caught by -Wconversion).
See, for example, ex_getln.c:2462: `windgoto(cmdline_row, 0);`, where `windgoto()` is defined as having an `int` first parameter.

So, going through that messy process would be the only way to do it correctly, if we really wanted to.
But we don't need that. Because, in fact, an int is enough for that variable (you won't have a screen with more than 2Glines). That is also true for `Rows` and `p_ch` (an `int` would be enough for them). But those *must* be `long` (because those are user options, and the way options are implemented by now, all numeric options have to be the same type, which turns out to be `long`). I'm in fact studying possible refactorings of this, to eliminate `long` from the codebase.

Then, how to remove the warning to let us enable -Wconversion?
In short, a safe cast.
This is, check that the substraction is within `int`bounds, and cast to int:
```
  assert(Rows - p_ch >= INT_MIN && Rows - p_ch <= INT_MAX);
  cmdline_row = (int)(Rows - p_ch);
```
In the general case, that wouldn't be completely correct yet, as the substraction could overflow/underflow. So, we'd have to check that substraction doesn't overflow as long, and result is within int bounds.
But we don't need that in this case, as we have extra knowledge (both Rows and p_ch must be positive, and Rows >= p_ch). So, introducing all that extra knowledge in the assert (which also helps us validate our assumptions about the code), it would remain:
```
  assert(p_ch >= 0 && Rows >= p_ch &&  Rows - p_ch <= INT_MAX);
  cmdline_row = (int)(Rows - p_ch);
```
It could still be improved, because it seems that Rows must be between 2 and 1000, and p_ch between 1 and Rows (exclusive). But that could be overkill, as limitation about Rows could be removed. Once both substraction and cast are safe, it's enough for me.

As you see, type refactoring is a tricky thing. Please be extra-picky when dealing with it. ;-)
Hope that helps.